core/src/main/scala/kafka/admin/LeaderElectionCommand.scala

```diff
@@ -0,0 +1,235 @@
+      val props = Option(commandOptions.options.valueOf(commandOptions.adminClientConfig)).map { config =>
+        Utils.loadProps(config)
+      }.getOrElse(new Properties())
+
+      props.setProperty(
+        AdminClientConfig.BOOTSTRAP_SERVERS_CONFIG,
+        commandOptions.options.valueOf(commandOptions.bootstrapServer)
+      )
+      props.setProperty(AdminClientConfig.REQUEST_TIMEOUT_MS_CONFIG, timeout.toMillis.toString)
+
+      JAdminClient.create(props)
+    }
+
+    try {
+      electLeaders(adminClient, electionType, topicPartitions)
+    } finally {
+      adminClient.close()
+    }
+  }
+
+  private[this] def parseReplicaElectionData(jsonString: String): Set[TopicPartition] = {
+    Json.parseFull(jsonString) match {
+      case Some(js) =>
+        js.asJsonObject.get("partitions") match {
+          case Some(partitionsList) =>
+            val partitionsRaw = partitionsList.asJsonArray.iterator.map(_.asJsonObject)
+            val partitions = partitionsRaw.map { p =>
+              val topic = p("topic").to[String]
+              val partition = p("partition").to[Int]
+              new TopicPartition(topic, partition)
+            }.toBuffer
+            val duplicatePartitions = CoreUtils.duplicates(partitions)
+            if (duplicatePartitions.nonEmpty) {
+              throw new AdminOperationException(
+                s"Replica election data contains duplicate partitions: ${duplicatePartitions.mkString(",")}"
+              )
+            }
+            partitions.toSet
+          case None => throw new AdminOperationException("Replica election data is missing \"partition\" field")
+        }
+      case None => throw new AdminOperationException("Replica election data is empty")
+    }
+  }
+
+  private[this] def electLeaders(
+    client: JAdminClient,
+    electionType: ElectionType,
+    topicPartitions: Option[Set[TopicPartition]]
+  ): Unit = {
+    val partitions = topicPartitions.map(_.asJava).getOrElse(null)
+    debug(s"Calling AdminClient.electLeaders($electionType, $partitions)")
+    val result = client.electLeaders(electionType, partitions)
+
+    val attemptedPartitions = topicPartitions.getOrElse {
+      try {
+        result.partitions().get.asScala
+      } catch {
+          case e: ExecutionException =>
+            val cause = e.getCause
+            if (cause.isInstanceOf[TimeoutException]) {
+              println("Timeout waiting for election results")
+            }
+            throw new AdminCommandFailedException("Timeout waiting for election results", cause)
+          case e: Throwable =>
+            println("Error while making request")
+            e.printStackTrace()
+            return
+      }
+    }
+
+    val (failed, succeeded) = attemptedPartitions
+      .map(topicPartition => topicPartition -> result.partitionResult(topicPartition))
+      .partition { case (_, partitionResult) => completedExceptionally(partitionResult) }
+
+    if (!succeeded.isEmpty) {
```

Comment:
Unless I'm missing something, this doesn't work for the weird `partitions()` semantics, which returns successfully for all partitions which are specified explicitly regardless of the election result.

Reply:
This all changed thanks to the much simpler API for `ElectLeadersResult`. Let me know if you still see logical errors.